src/irc/core/irc.c

```diff
@@ -49,55 +49,87 @@ static void strip_params_colon(char *const);
 void irc_send_cmd_full(IRC_SERVER_REC *server, const char *cmd,
 		       int send_now, int immediate, int raw)
 {
-	char str[513];
+	GString *str;
 	int len;
+	gboolean server_supports_tag;
 
 	g_return_if_fail(server != NULL);
 	g_return_if_fail(cmd != NULL);
 
 	if (server->connection_lost)
 		return;
 
-	len = strlen(cmd);
+	str = g_string_sized_new(MAX_IRC_USER_TAGS_LEN + 2 /* `@'+SPACE */ +
+				 server->max_message_len + 2 /* CR+LF */ + 1 /* `\0' */);
+
 	if (server->cmdcount == 0)
 		irc_servers_start_cmd_timeout();
 	server->cmdcount++;
 
 	if (!raw) {
+		const char *tmp = cmd;
+
+		server_supports_tag = server->cap_supported != NULL &&
+			g_hash_table_lookup_extended(server->cap_supported, CAP_MESSAGE_TAGS, NULL, NULL);
```

Comment:
You can have a message-tags section without the message tags cap being enabled, for example if you request the server-time, batch or account-tag caps. I think this entire check could just go, and we should handle the tags section unconditionally.

Reply:
Wait shit this is when sending. This is why i should use the 'start a review' thing to hide my shame when I realize things seconds after sending a comment.

My comment might still apply though. Need to keep reading.

Reply:
Yeah nevermind this entire thing is okay actually. It's enforcing the message-tags specific lengths for the tag section only, and doing tag-specific truncation which isn't a thing without the cap. Nevermind.

Reply:
:-)